include/seqan3/alignment/pairwise/alignment_algorithm.hpp

```diff
@@ -0,0 +1,168 @@
+
+        // Initialize cache variables to keep frequently used variables close to the CPU registers.
+        auto cache = this->setup_cache(get<align_cfg::gap>(*cfg_ptr).value);
+        // auto cache = this->setup_cache(gap_scheme{gap_score{-1}, gap_open_score{-10}});
+
+        initialize(cache);
+
+        compute(first_batch, second_batch, cache);
+
+        using result_t = typename select_result_type<first_batch_t, second_batch_t, config_t>::type;
+        result_t res{};
+
+        // Choose what needs to be computed.
+        if constexpr (config_t::template exists<align_cfg::result<detail::with_score_type>>())
+        {
+            auto last_col = this->active_column();
+            res.score = get<0>(*(std::ranges::end(last_col) - 1));
+        }
+        return align_result{res};
+    }
+
+protected:
+
+    /*!\brief Initializes the first column of the dynamic programming matrix.
+     * \tparam         cache_t The cache type.
+     * \param[in,out]  cache   The cache holding hot variables.
+     */
+    template <typename cache_t>
+    void initialize(cache_t & cache)
+    {
+        auto col = this->active_column();
+
+        this->init_origin_cell(*std::ranges::begin(col), cache);
+
+        ranges::for_each(col | ranges::view::drop_exactly(1), [&cache, this](auto & cell)
+        {
+            this->init_column_cell(cell, cache);
+        });
+    }
+
+    /*!\brief Compute the alignment by iterating over the dynamic programming matrix in a column wise manner.
+     * \tparam        first_batch_t  The type of the first sequence (or packed sequences).
+     * \tparam        second_batch_t The type of the second sequence (or packed sequences).
+     * \tparam        cache_t        The type of the cache.
+     * \param[in]     first_batch    The first sequence.
+     * \param[in]     second_batch   The second sequence.
+     * \param[in,out] cache          The cache holding hot variables.
+     */
+    template <typename first_batch_t,
+              typename second_batch_t,
+              typename cache_t>
+    void compute(first_batch_t const & first_batch,
+                 second_batch_t const & second_batch,
+                 cache_t & cache)
+    {
+        auto const & score_scheme = get<align_cfg::scoring>(*cfg_ptr).value;
+        ranges::for_each(first_batch, [&, this](auto seq1_value)
+        {
+            auto col = this->active_column();
+            this->init_row_cell(*std::ranges::begin(col), cache);
+
+            auto second_batch_it = std::ranges::begin(second_batch);
+            ranges::for_each(col | ranges::view::drop_exactly(1), [&, this] (auto & cell)
+            {
+                this->compute_cell(cell, cache, score_scheme.score(seq1_value, *second_batch_it));
+                ++second_batch_it;
+            });
+        });
+    }
+
+private:
+
+    //!\brief The alignment configuration stored on the heap.
+    std::shared_ptr<config_t> cfg_ptr{};
```

Comment:
why is the config a member of the type, but the cache a local variable? Are any of these things called in parallel?

Reply:
It's just that then the burder is on the user to take care of the cache, right?

Reply:
No? The user never sees nor touches any of this. This is all implementation specific stuff for library developers only.